src/librustc/ty/layout.rs

```diff
@@ -1442,7 +1442,31 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
         debug!("generator layout ({:?}): {:#?}", ty, layout);
         Ok(layout)
     }
+}
 
+fn recompute_memory_index(offsets: &[Size], fields: &[TyLayout<'_>]) -> Vec<u32> {
+    debug!("recompute_memory_index({:?})", offsets);
+    debug!("fields = {:#?}", fields);
+    let mut inverse_index = (0..offsets.len() as u32).collect::<Vec<_>>();
+    inverse_index.sort_unstable_by_key(|i| {
+        // Place ZSTs before other fields at the same offset.
+        // Codegen expects this.
```

Comment:
I'm confused, I don't see why this would be guaranteed for even a plain `struct`?

Reply:
Nevermind, I think what you mean is simply that you can't put fields out of order - a non-ZST can't be followed by a field of the same offset, because the non-ZST *had advanced* the offset.
I'll try a much simpler fix, to confirm my suspicions.

Reply:
Yep, that's what I mean.